test/gtest/core/test_reduce.cc

```diff
@@ -0,0 +1,247 @@
+                   UccCollCtxVec &ctxs)
+    {
+        ctxs.resize(nprocs);
+        for (int r = 0; r < nprocs; r++) {
+            ucc_coll_args_t *coll = (ucc_coll_args_t*)
+                    calloc(1, sizeof(ucc_coll_args_t));
+
+            ctxs[r] = (gtest_ucc_coll_ctx_t*)calloc(1,
+                          sizeof(gtest_ucc_coll_ctx_t));
+            ctxs[r]->args = coll;
+
+            coll->mask = UCC_COLL_ARGS_FIELD_PREDEFINED_REDUCTIONS;
+            coll->coll_type = UCC_COLL_TYPE_REDUCE;
+            coll->reduce.predefined_op = T::redop;
+            coll->root = root;
+            coll->src.info.mem_type = mem_type;
+            coll->src.info.count   = (ucc_count_t)count;
+            coll->src.info.datatype = dt;
+
+            ctxs[r]->init_buf = ucc_malloc(ucc_dt_size(dt) * count,
+                                                        "init buf");
+            EXPECT_NE(ctxs[r]->init_buf, nullptr);
+            for (int i = 0; i < count; i++) {
+                typename T::type * ptr;
+                ptr = (typename T::type *)ctxs[r]->init_buf;
+                /* need to limit the init value so that "prod" operation
+                   would not grow too large. We have teams up to 16 procs
+                   in gtest, this would result in prod ~2**48 */
+                ptr[i] = (typename T::type)((i + r + 1) % 8);
+            }
+
+            if (r == root) {
+			    coll->dst.info.mem_type = mem_type;
+			    coll->dst.info.count   = (ucc_count_t)count;
+			    coll->dst.info.datatype = dt;
+			    UCC_CHECK(ucc_mc_alloc(&ctxs[r]->dst_mc_header,
+			                           ucc_dt_size(dt) * count, mem_type));
+			    coll->dst.info.buffer = ctxs[r]->dst_mc_header->addr;
+			    if (inplace) {
+			        UCC_CHECK(ucc_mc_memcpy(coll->dst.info.buffer,
+			    		  ctxs[r]->init_buf, ucc_dt_size(dt) * count, mem_type,
+			              UCC_MEMORY_TYPE_HOST));
+			    }
+            }
+            if (inplace) {
+                coll->mask  |= UCC_COLL_ARGS_FIELD_FLAGS;
+                coll->flags |= UCC_COLL_ARGS_FLAG_IN_PLACE;
+
+            }
+            if (r != root || !inplace) {
+                UCC_CHECK(ucc_mc_alloc(&ctxs[r]->src_mc_header,
+                                       ucc_dt_size(dt) * count, mem_type));
+                coll->src.info.buffer = ctxs[r]->src_mc_header->addr;
+                UCC_CHECK(ucc_mc_memcpy(coll->src.info.buffer,
+                                        ctxs[r]->init_buf,
+                                        ucc_dt_size(dt) * count, mem_type,
+                                        UCC_MEMORY_TYPE_HOST));
+            }
+        }
+    }
+    void data_fini(UccCollCtxVec ctxs) {
+    	for (auto r = 0; r < ctxs.size(); r++) {
+            ucc_coll_args_t* coll = ctxs[r]->args;
+            if (r == root) {
+                UCC_CHECK(ucc_mc_free(ctxs[r]->dst_mc_header));
+            }
+            if (coll->src.info.buffer) { /* no inplace */
```

Comment:
why not the same condition as alloc?

Reply:
You're right, I now changed it to be aligned with alloc.